File: src/conventions/pipeline/evidence-bundle.ts
/**
 * Evidence types for decision program representations.
 *
 * ConditionResult is the shared base for all "did a condition pass?"
 * shapes across evidence, teaching, and pipeline types.
 *
 * ConditionEvidence extends it as the canonical evidence type for
 * pipeline provenance and evidence bundles.
 */

/** Role a condition plays in the evaluation pipeline. */
export type ConditionRole = "semantic" | "inferential" | "pedagogical" | "routing";

/** Base condition evaluation result — shared across evidence, teaching, and pipeline types. */
export interface ConditionResult {
  readonly conditionId?: string;
  readonly factId?: string;
  readonly satisfied: boolean;
  readonly description?: string;
  readonly observedValue?: unknown;
  readonly threshold?: unknown;
  readonly conditionRole?: ConditionRole;
}

/** Evidence for a single condition evaluation. */
export interface ConditionEvidence extends ConditionResult {
  readonly conditionId: string;
  /** Parameters for parameterized facts (e.g. `{ suit: "hearts" }` for `suitLength(hearts)`). */
  readonly params?: Readonly<Record<string, unknown>>;
}

/** Evidence for why a meaning was rejected. */
export interface RejectionEvidence {
  readonly meaningId: string;
  readonly failedConditions: readonly ConditionEvidence[];
  readonly moduleId: string;
  /** Conditions whose failure supports negative inference (invertInference). Optional for backward compat. */
  readonly negatableFailures?: readonly ConditionEvidence[];
}

/** Evidence for an alternative meaning that was considered. */
export interface AlternativeEvidence {
  readonly meaningId: string;
  readonly call: string;
  readonly ranking: { readonly band: string; readonly specificity: number };
  readonly reason: string;
  /** Per-condition delta vs. matched meaning (spec: Agreement Module IR). Optional for backward compat. */
  readonly conditionDelta?: readonly ConditionEvidence[];
}

/** The spec's evidence contract for decision program representations. */
export interface EvidenceBundle {
  readonly matched: { meaningId: string; satisfiedConditions: ConditionEvidence[] } | null;
  readonly rejected: readonly RejectionEvidence[];
  readonly alternatives: readonly AlternativeEvidence[];
  readonly exhaustive: boolean;
  readonly fallbackReached: boolean;
}

File: crates/bridge-conventions/src/pipeline/evidence_bundle.rs
//! Evidence types for decision program representations.
//!
//! Mirrors TS types from `pipeline/evidence-bundle.ts`.

use serde::{Deserialize, Serialize};


/// Evidence for a single condition evaluation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConditionEvidence {
    pub condition_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fact_id: Option<String>,
    pub satisfied: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub observed_value: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub threshold: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<std::collections::HashMap<String, serde_json::Value>>,
}

/// Evidence for why a meaning was rejected.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RejectionEvidence {
    pub meaning_id: String,
    pub failed_conditions: Vec<ConditionEvidence>,
    pub module_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub negatable_failures: Option<Vec<ConditionEvidence>>,
}

/// Evidence for an alternative meaning that was considered.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AlternativeEvidence {
    pub meaning_id: String,
    pub call: String,
    pub ranking: AlternativeRanking,
    pub reason: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub condition_delta: Option<Vec<ConditionEvidence>>,
}

/// Ranking info for alternative evidence.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AlternativeRanking {
    pub band: String,
    pub specificity: f64,
}

/// The evidence contract for decision program representations.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceBundle {
    pub matched: Option<MatchedEvidence>,
    pub rejected: Vec<RejectionEvidence>,
    pub alternatives: Vec<AlternativeEvidence>,
    pub exhaustive: bool,
    pub fallback_reached: bool,
}

/// Evidence for the matched meaning.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MatchedEvidence {
    pub meaning_id: String,
    pub satisfied_conditions: Vec<ConditionEvidence>,
}
